docs(messaging): illustrate how to use `getInitialMessage()` & `onMessageOpenedApp` (#10473)
firebase/flutterfire: docs/cloud-messaging/first-message.md

devices, along with data for "impressions" (notifications seen by users) for
Android apps.

## Handling interaction

> The default behavior for notifications (e.g. by pressing them) on both Android & iOS is to open the application. If the application is terminated,

Reviewer: I still don't think this parenthetical is at all required.  Two points:

-- Even setting aside non-sighted usage, "pressing" is nonstandard;  I've heard many references to "tapping" on mobile GUIs, but I strongly prefer the agnostic "selecting" for such cases.  

-- "e.g." is discourage by the Google developer style guide, here: https://developers.google.com/style/abbreviations#dont-use

In any case, I don't think this parenthetical adds anything.  Without it, the sentence seems perfectly clear to me. The only clarification I can think of might be to clarify that this behavior is initiated by the interaction, which I think can be safely assumed.  But if you and Kevin feel it is necessary to clarify, how about "When users select a notification, the default behavior on both . . . ."  ?

Reviewer: SGTM.

devices, along with data for "impressions" (notifications seen by users) for
Android apps.

> ## Handling Interaction

Reviewer: No need to capitalize "interaction."

> Since notifications are a visible cue, it is common for users to interact with it (by pressing them).

Reviewer: I *think* that some special-accessibility interfaces allow interaction with non-visible (specifically audio-rendered) notification widgets, so this entire sentence may be best to omit.

Suggest starting with just "Default notification behavior on both Android and . . . "

The default behavior on both Android & iOS is to open the application. If the application is terminated

> it will be started, if it is in the background it will be brought to the foreground.

Reviewer: Add ", and" or just a period to avoid comma splice.

> Depending on the content of a notification, you may wish to handle the users interaction when the application

Reviewer: "user's"

> opens. For example, if a new chat message is sent via a notification and the user presses it, you may want to

Reviewer: Suggest "selects" instead of "presses."

 open the specific conversation when the application opens.

The `firebase-messaging` package provides two ways to handle this interaction:

1. `getInitialMessage()`: If the application is opened from a terminated state a `Future` containing a
    `RemoteMessage` will be returned. Once consumed, the `RemoteMessage` will be removed.
2. `onMessageOpenedApp`: A `Stream` which posts a `RemoteMessage` when the application is opened from a
    background state.

> It is recommended that both scenarios are handled to ensure a smooth UX for your users. The code example

Reviewer: Suggest "experience" instead of  "UX"

> Since notifications are a visible cue, it is common for users to interact with it (by pressing them).
> The default behavior on both Android & iOS is to open the application. If the application is terminated
> it will be started, if it is in the background it will be brought to the foreground.
> 
> Depending on the content of a notification, you may wish to handle the users interaction when the application
> opens. For example, if a new chat message is sent via a notification and the user presses it, you may want to
>  open the specific conversation when the application opens.
> 
> The `firebase-messaging` package provides two ways to handle this interaction:
> 
> 1. `getInitialMessage()`: If the application is opened from a terminated state a `Future` containing a
>     `RemoteMessage` will be returned. Once consumed, the `RemoteMessage` will be removed.
> 2. `onMessageOpenedApp`: A `Stream` which posts a `RemoteMessage` when the application is opened from a
>     background state.
> 
> It is recommended that both scenarios are handled to ensure a smooth UX for your users. The code example
> below outlines how this can be achieved:

Reviewer 2: ```suggestion
Since notifications are a visible cue, it is common for users to interact with them (by pressing them).
The default behavior on both Android & iOS is to open the application. If the application is terminated,
it will be started, and if it is in the background, it will be brought to the foreground.

Depending on the content of a notification, you may wish to handle the user's interaction when the application
opens. For example, if a new chat message is sent using a notification and the user presses it, you may want to
 open the specific conversation when the application opens.

The `firebase-messaging` package provides two ways to handle this interaction:

1. `getInitialMessage()`: If the application is opened from a terminated state, this method returns a `Future` containing a `RemoteMessage`. Once consumed, the `RemoteMessage` will be removed.
2. `onMessageOpenedApp`: A `Stream` which posts a `RemoteMessage` when the application is opened from a
    background state.

To ensure a smooth UX for your users, you should handle both scenarios. The code example
below outlines how this can be achieved:
```

```dart
class Application extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => _Application();
}

class _Application extends State<Application> {

>   // It is assumed that all messages contain a data field with the key 'type'

Reviewer 2: Can a developer assume this or is this an assumption for the sake of illustration?

Author: assumption for the sake of illustration.
